### PR TITLE
Determine fuzziness of bleve indexer by keyword length

### PR DESCRIPTION
**Pull request by 6543**
also bleve did match on fuzzy search and the other way around. this also fix that bug.

### CODE REVIEW

**Comment by 6543** [comment]
PS: if we dont want that simple enhancement in v1.22.0 I can extract the bugfix, but I would prefere to simply have this fixed and enhanced ;)

**Comment by 6543** [comment]
@lunny :point_up:

**Review by silverwind — APPROVED** [review]

**Review by denyskon — APPROVED** [review]

**Review by delvh — APPROVED** [review]

File: tests/integration/repo_search_test.go [hunk 1]
```diff
@@ -58,6 +61,17 @@ func testSearch(t *testing.T, url string, expected []string) {
 	assert.EqualValues(t, expected, filenames)
 }
 
-func executeIndexer(t *testing.T, repo *repo_model.Repository, op func(*repo_model.Repository)) {
-	op(repo)
+func executeIndexer(t *testing.T, repo *repo_model.Repository) {
+	code_indexer.UpdateRepoIndexer(repo)
+
+	for {
```
Comment (lunny) [inline tests/integration/repo_search_test.go]: Why we need these code?

**Review by lunny — DISMISSED** [review]
per https://github.com/go-gitea/gitea/pull/29706/files#r1525990208

File: tests/integration/repo_search_test.go (hunk 1 above)
Comment (6543) [reply tests/integration/repo_search_test.go]: as the code before just throw the repo in the indexer queue and we never wait for it to be sure it got finished

that way the test is for sure deterministic

an alternative would be to just add a sleep and hope after it the indexer is done (witch normally hapens quite fast (witch is why the test did work in the past))

Comment (lunny) [reply tests/integration/repo_search_test.go]: I think the queues on tests are all synchronous.

Comment (6543) [reply tests/integration/repo_search_test.go]: In this case i can rm that ...

**Comment by 6543** [comment]
@lunny we now ger test errors ... :/

File: tests/integration/repo_search_test.go [hunk 2]
```diff
@@ -32,7 +32,7 @@ func TestSearchRepo(t *testing.T) {
 	repo, err := repo_model.GetRepositoryByOwnerAndName(db.DefaultContext, "user2", "repo1")
 	assert.NoError(t, err)
 
-	executeIndexer(t, repo, code_indexer.UpdateRepoIndexer)
+	code_indexer.UpdateRepoIndexer(repo)
```
Comment (6543) [inline tests/integration/repo_search_test.go:35]:
```suggestion
	code_indexer.UpdateRepoIndexer(repo)
	time.Sleep(1000)
```

File: tests/integration/repo_search_test.go [hunk 3]
```diff
@@ -42,12 +42,14 @@ func TestSearchRepo(t *testing.T) {
 	repo, err = repo_model.GetRepositoryByOwnerAndName(db.DefaultContext, "user2", "glob")
 	assert.NoError(t, err)
 
-	executeIndexer(t, repo, code_indexer.UpdateRepoIndexer)
+	code_indexer.UpdateRepoIndexer(repo)
```
Comment (6543) [inline tests/integration/repo_search_test.go:45]:
```suggestion
	code_indexer.UpdateRepoIndexer(repo)
	time.Sleep(1000)
```

File: tests/integration/repo_search_test.go [hunk 4]
```diff
@@ -32,7 +33,8 @@ func TestSearchRepo(t *testing.T) {
 	repo, err := repo_model.GetRepositoryByOwnerAndName(db.DefaultContext, "user2", "repo1")
 	assert.NoError(t, err)
 
-	executeIndexer(t, repo, code_indexer.UpdateRepoIndexer)
+	code_indexer.UpdateRepoIndexer(repo)
+	time.Sleep(1000)
```
Comment (6543) [inline tests/integration/repo_search_test.go]:
```suggestion
	time.Sleep(100 * time.Millisecond)
```

File: tests/integration/repo_search_test.go [hunk 5]
```diff
@@ -42,12 +44,15 @@ func TestSearchRepo(t *testing.T) {
 	repo, err = repo_model.GetRepositoryByOwnerAndName(db.DefaultContext, "user2", "glob")
 	assert.NoError(t, err)
 
-	executeIndexer(t, repo, code_indexer.UpdateRepoIndexer)
+	code_indexer.UpdateRepoIndexer(repo)
+	time.Sleep(1000)
```
Comment (6543) [inline tests/integration/repo_search_test.go]:
```suggestion
	time.Sleep(100 * time.Millisecond)
```

File: tests/integration/repo_search_test.go [hunk 6]
```diff
@@ -42,12 +44,15 @@ func TestSearchRepo(t *testing.T) {
...
+	code_indexer.UpdateRepoIndexer(repo)
+	time.Sleep(100 * time.Millisecond)
 
 	testSearch(t, "/user2/glob/search?q=loren&page=1", []string{"a.txt"})
-	testSearch(t, "/user2/glob/search?q=file3&page=1", []string{"x/b.txt"})
-	testSearch(t, "/user2/glob/search?q=file4&page=1", []string{})
-	testSearch(t, "/user2/glob/search?q=file5&page=1", []string{})
+	testSearch(t, "/user2/glob/search?q=loren&page=1&t=match", []string{"a.txt"})
+	testSearch(t, "/user2/glob/search?q=file3&page=1", []string{"x/b.txt", "a.txt"})
+	testSearch(t, "/user2/glob/search?q=file3&page=1&t=match", []string{"x/b.txt"})
```
Comment (6543) [inline tests/integration/repo_search_test.go]:
```suggestion
	testSearch(t, "/user2/glob/search?q=file3&page=1&t=match", []string{"x/b.txt", "a.txt"})
```

File: tests/integration/repo_search_test.go [hunk 7]
```diff
@@ -42,12 +44,15 @@ func TestSearchRepo(t *testing.T) {
 	repo, err = repo_model.GetRepositoryByOwnerAndName(db.DefaultContext, "user2", "glob")
 	assert.NoError(t, err)
 
-	executeIndexer(t, repo, code_indexer.UpdateRepoIndexer)
+	code_indexer.UpdateRepoIndexer(repo)
+	time.Sleep(100 * time.Millisecond)
```
Comment (6543) [inline tests/integration/repo_search_test.go]:
```suggestion
```

File: tests/integration/repo_search_test.go [hunk 8]
```diff
@@ -32,7 +33,8 @@ func TestSearchRepo(t *testing.T) {
 	repo, err := repo_model.GetRepositoryByOwnerAndName(db.DefaultContext, "user2", "repo1")
 	assert.NoError(t, err)
 
-	executeIndexer(t, repo, code_indexer.UpdateRepoIndexer)
+	code_indexer.UpdateRepoIndexer(repo)
+	time.Sleep(100 * time.Millisecond)
```
Comment (6543) [inline tests/integration/repo_search_test.go]:
```suggestion
```

File: tests/integration/repo_search_test.go [hunk 9]
```diff
@@ -6,6 +6,7 @@ package integration
 import (
 	"net/http"
 	"testing"
+	"time"
```
Comment (6543) [inline tests/integration/repo_search_test.go]:
```suggestion
```

File: tests/integration/repo_search_test.go [hunk 10]
```diff
@@ -42,12 +42,14 @@ func TestSearchRepo(t *testing.T) {
...
+	code_indexer.UpdateRepoIndexer(repo)
 
 	testSearch(t, "/user2/glob/search?q=loren&page=1", []string{"a.txt"})
-	testSearch(t, "/user2/glob/search?q=file3&page=1", []string{"x/b.txt"})
-	testSearch(t, "/user2/glob/search?q=file4&page=1", []string{})
-	testSearch(t, "/user2/glob/search?q=file5&page=1", []string{})
+	testSearch(t, "/user2/glob/search?q=loren&page=1&t=match", []string{"a.txt"})
+	testSearch(t, "/user2/glob/search?q=file3&page=1", []string{"x/b.txt", "a.txt"})
+	testSearch(t, "/user2/glob/search?q=file3&page=1&t=match", []string{"x/b.txt", "a.txt"})
+	testSearch(t, "/user2/glob/search?q=file4&page=1&t=match", []string{})
```
Comment (6543) [inline tests/integration/repo_search_test.go]:
```suggestion
	testSearch(t, "/user2/glob/search?q=file4&page=1&t=match", []string{"x/b.txt", "a.txt"})
```

File: tests/integration/repo_search_test.go [hunk 11]
```diff
@@ -42,12 +42,14 @@ func TestSearchRepo(t *testing.T) {
...
+	testSearch(t, "/user2/glob/search?q=file5&page=1&t=match", []string{})
```
Comment (6543) [inline tests/integration/repo_search_test.go]:
```suggestion
	testSearch(t, "/user2/glob/search?q=file5&page=1&t=match", []string{"x/b.txt", "a.txt"})
```

**Comment by 6543** [comment]
☝️ fixing via mobile sucks :/

**Comment by KN4CK3R** [comment]
Then don't do it. I don't need 20 notifications. 🙂

**Comment by wxiaoguang** [comment]
There are a lot of bug reports for 1.22 Search Functionality Issues with Bleeve Engine #31565 

The fuzzy related PRs are made by you, would you take a look?

**Pull request merged**